lib/Sema/SemaBounds.cpp

```diff
@@ -2145,11 +2153,143 @@ namespace {
+  // CFG elements.  (CFG elements are the components of basic blocks).  When a
+  // CFG is constructed, subexpressions of top-level expressions may be placed
+  // in separate CFG elements.  This is done for subexpressions of expressions
+  // with control-flow, for example. When checking bounds declarations, we want
+  // to process a subexpression with its enclosing expression. We want to
+  // ignore CFG elements that are substatements of other CFG elements.
+   void FindNestedElements(StmtSet &NestedStmts) {
+      // Create the set of top-level CFG elements.
+      StmtSet TopLevelElems;
+      CFG::iterator Iter = Cfg->begin();
+      CFG::iterator IterEnd = Cfg->end();
+      for ( ; Iter != IterEnd; ++Iter) {
+        const CFGBlock *Block = *Iter;
+        CFGBlock::const_iterator ElemIter = Block->begin();
+        CFGBlock::const_iterator ElemEnd = Block->end();
+        for (; ElemIter != ElemEnd; ++ElemIter) {
+          CFGElement Elem = *ElemIter;
+          if (Elem.getKind() == CFGElement::Statement) {
+            CFGStmt CS = Elem.castAs<CFGStmt>();
+            const Stmt *S = CS.getStmt();
+            TopLevelElems.insert(S);
+          }
+        }
+      }
+
+      // Create the set of top-level elements that are subexpressions
+      // of other top-level elements.
+      for (Iter = Cfg->begin(); Iter != IterEnd; ++Iter) {
+        const CFGBlock *Block = *Iter;
+        CFGBlock::const_iterator ElemIter = Block->begin();
+        CFGBlock::const_iterator ElemEnd = Block->end();
+        for (; ElemIter != ElemEnd; ++ElemIter) {
+          CFGElement Elem = *ElemIter;
+          if (Elem.getKind() == CFGElement::Statement) {
+            CFGStmt CS = Elem.castAs<CFGStmt>();
+            const Stmt *S = CS.getStmt();
+            MarkNested(S, NestedStmts, TopLevelElems);
+          }
+        }
+      }
+   }
+
+   // Walk the CFG, traversing basic blocks in reverse post-oder.
+   // For each element of a block, check bounds declarations.  Skip
+   // CFG elements that are subexpressions of other CFG elements.
+   void TraverseCFG() {
+     assert(Cfg && "expected CFG to exist");
+#if TRACE_CFG
+     llvm::outs() << "Dumping AST";
+     Body->dump(llvm::outs());
+     llvm::outs() << "Dumping CFG:\n";
+     Cfg->print(llvm::outs(), S.getLangOpts(), true);
+     llvm::outs() << "Traversing CFG:\n";
+#endif
+     StmtSet NestedElements;
+     FindNestedElements(NestedElements);
+     StmtSet CheckedStmts;
+     IdentifyChecked(Body, CheckedStmts, false);
+     PostOrderCFGView POView = PostOrderCFGView(Cfg);
+     PostOrderCFGView::iterator PO_Iter= POView.begin();
+     PostOrderCFGView::iterator PO_End = POView.end();
+     for ( ; PO_Iter != PO_End; ++PO_Iter) {
+       const CFGBlock *Block = *PO_Iter;
+       CFGBlock::const_iterator ElemIter = Block->begin();
+       CFGBlock::const_iterator ElemEnd = Block->end();
+       for (; ElemIter != ElemEnd; ++ElemIter) {
```

Comment:
Same as line 2206, I thinkg using "for (const auto& PO_iter : POView)" can simplify the code